src/Adapter/ContextStateManager.php

```diff
@@ -153,9 +153,30 @@ public function setCustomer(?Customer $customer): self
      */
     public function restoreContext(): self
     {
-        foreach ($this->savedContextFields as $fieldName => $contextValue) {
+        $currentStashIndex = count($this->stashedContextsFields) - 1;
+        foreach ($this->stashedContextsFields[$currentStashIndex] as $fieldName => $contextValue) {
             $this->restoreContextField($fieldName);
         }
+        $this->popContext();
```

Comment:
Why popContext when you restore it? :thinking:

Reply:
The words comes from `git stash` / `git stash pop`

At first I added a public method `popContext` that called `restoreContext` and then pop the internal stack But the idea of this modification was to limit the modification so that this class doesn't change too much in case it is merged for the patch version only (to avoid BC break)

So I don't want to change the initial behaviour when `stachContext` is not used, that point is managed

I could keep the `popContext` method public to emphasize the fact that I perform two actions, but since you usually `restoreContext` once you don't need to override it any more it makes sense that it au pops the previous stashed context This was the only endpoint that you need when you're done using the state manager is always the same `restoreContext`

Reply:
`git stash pop` is applying changes to the current codebase. Here, you're only use `array_pop`, without using the result of the call. I'm more like a `git stash drop`.
Then if you want to insert the `stash` logic, maybe all methods must include the `stash` keyword in the functions' name?

Reply:
"Removes the last saved stashed context"

then why not just `removeLastStashedContext`?

Reply:
The git stash influence is probably not adapted, that's what I had in mind at the moment I coded it but there's room for improvement!
`removeLastStashedContext` seems a bit long 🤔

When I talked with @Progi1984 he mentioned a notion of "breakpoint" maybe that's a lead.. Or something inspired on DB transaction/rollback

Reply:
I'm a fun of code which you can literally read, if there's no shorter and descriptive function name as an alternative then I'd go for it

`rollbackContext` also sounds better

Reply:
@kpodemski I kept your idea for `removeLastSavedContext` (stashed was replaced with saved) I forgot it was a private function so it's better to be clear I tend to like small names for public function (when possible)
But considering the complexity of this service anyway it's safer to have clear names everywhere

Reply:
> I'm a fan of code which you can literally read

Me too ! But when I say that, my colleagues tell me I'm a Java developer 😞